core/src/test/java/org/apache/brooklyn/core/mgmt/internal/EntityExecutionManagerTest.java

```diff
@@ -69,32 +71,47 @@
 import com.google.common.util.concurrent.Callables;
 
 /** Includes many tests for {@link BrooklynGarbageCollector} */
+@Test
 public class EntityExecutionManagerTest extends BrooklynAppUnitTestSupport {
     
     private static final Logger LOG = LoggerFactory.getLogger(EntityExecutionManagerTest.class);
-    
-    @Test
+
     public void testOnDoneCallback() throws InterruptedException {
         ExecutionManager em = mgmt.getExecutionManager();
         BasicExecutionManager bem = (BasicExecutionManager)em;
         final Map<Task<?>,Duration> completedTasks = MutableMap.of();
-        final Semaphore sema4 = new Semaphore(-1);
+        final CountDownLatch latch = new CountDownLatch(2);
         bem.addListener(new ExecutionListener() {
             @Override
             public void onTaskDone(Task<?> task) {
                 Assert.assertTrue(task.isDone());
-                Assert.assertEquals(task.getUnchecked(), "foo");
-                completedTasks.put(task, Duration.sinceUtc(task.getEndTimeUtc()));
-                sema4.release();
+                Object result = task.getUnchecked();
+                if (result != null && result.equals("foo")) {
+                    synchronized (completedTasks) {
+                        completedTasks.put(task, Duration.sinceUtc(task.getEndTimeUtc()));
+                    }
+                    latch.countDown();
+                }
             }
         });
-        Task<String> t1 = em.submit( Tasks.<String>builder().displayName("t1").dynamic(false).body(Callables.returning("foo")).build() );
-        t1.getUnchecked();
-        Task<String> t2 = em.submit( Tasks.<String>builder().displayName("t2").dynamic(false).body(Callables.returning("foo")).build() );
-        sema4.acquire();
-        Assert.assertEquals(completedTasks.size(), 2, "completed tasks are: "+completedTasks);
-        completedTasks.get(t1).isShorterThan(Duration.TEN_SECONDS);
-        completedTasks.get(t2).isShorterThan(Duration.TEN_SECONDS);
+        Task<String> t1 = em.submit(
+            Tasks.<String>builder()
+                .displayName("t1")
+                .dynamic(false)
+                .body(Callables.returning("foo"))
+                .build());
+        Task<String> t2 = em.submit(
+            Tasks.<String>builder()
+                .displayName("t2")
+                .dynamic(false)
+                .body(Callables.returning("foo"))
+                .build());
+        latch.await(Asserts.DEFAULT_LONG_TIMEOUT.toMilliseconds(), TimeUnit.MILLISECONDS);
```

Comment:
FYI This will return `false` if not satisfied within the given time. But that's fine. The assertion below will then fail.